### ISSUE TITLE
[Request] Add support for pg_repack extension on Heroku Postgres

### ISSUE THREAD

**Issue by jeffblake** [heroku/roadmap#123]
### Required Terms

- [X] I agree to follow this project's [Code of Conduct](https://github.com/heroku/roadmap/blob/main/CODE_OF_CONDUCT.md)
- [X] I have read and accept the [Salesforce Program Agreement](https://www.salesforce.com/company/program-agreement/)

### What service(s) is this request for?

Postgres

### Tell us about what you're trying to solve. What challenges are you facing?

```markdown
Title pretty much says it all. Without pg_repack, the only way to reclaim disk space is really `REINDEX INDEX CONCURRENTLY`, `VACUUM FULL`, and `pg:copy`, the latter two of which are pretty invasive and require downtime. `pg_repack` would be a big help! There is also `pg_squeeze

https://github.com/reorg/pg_repack
https://github.com/cybertec-postgresql/pg_squeeze
```

**Comment by zmalone** [comment]
This would be a huge improvement, at probably marginal cost. Right now, it's extremely hard to manage table bloat on Heroku, and other 3rd party Postgres extensions already exist on the platform.

**Comment by sasharevzin** [comment]
Or maybe pg_squeeze?

**Comment by 72L** [comment]
This is likely counter to Heroku's incentives.

For others here, what else can we do to reduce bloat? Is there a way to run a sequence of psql commands that would achieve something similar to what pg_repack does?

**Comment by sasharevzin** [comment]
> This is likely counter to Heroku's incentives.
> 
> For others here, what else can we do to reduce bloat? Is there a way to run a sequence of psql commands that would achieve something similar to what pg_repack does?

You might look at this project https://github.com/shayonj/pg-osc

**Comment by locofocos** [comment]
We're slowly facing scaling challenges with our Postgres database. Our multi-tenant software organizes our data by tenant_id, inside a single monolithic database.

Heroku supports declarative partitioning to help with this https://devcenter.heroku.com/articles/increasing-performance-of-large-tables-using-partitioning but the schema requirements (around unique constraints, PK constraints, FKs, etc) presented numerous challenges when we tried this.

We've tried running CLUSTER (for example, `CLUSTER VERBOSE products USING index_products_on_tenant_id`) and it made a massive improvement to our query performance in a stale database. But CLUSTER requires a long downtime, and our application can't tolerate that. 

`pg_repack` would allow us to run CLUSTER without a downtime. So by not supporting this, Heroku doesn't really offer the scaling we need at a competitive price point.

As it stands, AWS offers a Postgres database with support for pg_repack https://aws.amazon.com/blogs/database/remove-bloat-from-amazon-aurora-and-rds-for-postgresql-with-pg_repack/ so migrating off of Heroku seems like the only viable long-term option for my team.

**Comment by jbrown-heroku** [comment]
Thank you @jeffblake for raising this. 

We now have `pg_repack` roadmapped with our new Heroku database platform we are currently building. The release is targeted for early 2025.

**Comment by yigaldviri** [comment]
> Thank you [@jeffblake](https://github.com/jeffblake) for raising this.
> 
> We now have `pg_repack` roadmapped with our new Heroku database platform we are currently building. The release is targeted for early 2025.

Hi there, do you an estimation when will it be available?

**Comment by axlekb** [comment]
Hi heroku folks, we are now a few weeks from what most would consider the end of "early 2025".  This issue appears to still be in the "Researching" phase.  Is there any chance we could get a new target for when this might get worked on?

Really looking forward to using this feature!

**Comment by bambielli** [comment]
Agreed, could we get a status update on this from someone on the heroku team @jbrown-heroku ?

**Comment by jchatel** [comment]
Knock knock

**Comment by cesarmtzrdz** [comment]
![Image](https://github.com/user-attachments/assets/e4ae56a3-2f59-4c3a-ba5d-70f6f518af48)

all jokes aside, an update would be cool.

**Comment by jessehanley** [comment]
@jchatel funny to see you here. 

@jbrown-heroku any chance we can get pg_repack on the existing Heroku offering given the new platform isn't ready?

**Comment by kapoormanish** [comment]
@jbrown-heroku  Are there any further updates on this topic?

**Comment by aesnyder** [comment]
@jbrown-heroku any update here?

**Comment by danielricecodes** [comment]
Chiming in here.  I'm going to be looking for alternatives to Heroku Postgres if this issue isn't taken seriously.  I shouldn't have to upgrade from Standard 2 - and beyond - just because its impractical/impossible to run `VACUUM FULL`.  My app runs 24x7 and can't be put into maintenance mode for hours.  I need to be able to run `pg_repack` on my production database, and I can't see a viable way to do so that doesn't incur unacceptable and unknown amounts downtime for my client.

**Comment by aesnyder** [comment]
Yep, we are planning our exit strategy too.

**Comment by danielricecodes** [comment]
> Yep, we are planning our exit strategy too.

Same.  The solution to the exact problem everyone is having here is `pg_repack` and it appears impossible to practically use it on Heroku as-is.  `heroku pg:repack --table-name=users` <== thats what is needed here.

**Comment by cyril-sf** [comment]
From another issue #152 (and #336):

> As soon as we have the external pilot version ready, we will reach out. 
> As an update, we are currently running internal pilot. External pilot still requires a few months - I will keep this ambiguous at this time. I will update if we have a firmer date.

That was 2 months ago.

Not sure if our issue will be part of the pilot...

I'm guessing that someone learned the hard way to not provide release dates to customers 😅

**Comment by jbrown-heroku** [comment]
Sorry folks for the delay.
I understand the pain of not having `pg_repack`. We are still focused on building the new platform today, however, I will explore ways to add this to our current offerings if possible.

**Comment by danielricecodes** [comment]
> Yep, we are planning our exit strategy too.

FYI - if I find a way to efficiently reduce my database size, I'll let you know.  I might have to sit through a long `pg:copy` after running all the maintenance with a local copy of my production database first.

**Comment by jessehanley** [comment]
Not having this, not having logical replication, and constantly being told the new platform is "next month" is depressing.

**Comment by joshhepworth** [comment]
Any updates here? We're very interested in the availability of this and will need a solution soon. If we have to go with a route that involves downtime, we're likely to just migrate to a provider that has support for `pg_repack` instead during that downtime.

**Comment by jessehanley** [comment]
@jbrown-heroku what is the ETA on the new platform?

**Comment by jchatel** [comment]
Can anyone start listing alternative solutions here? At least it would be more helpful to people reading.
We are on Standard 7 and I'm not going into another 4h maintenance.
We decided to move away from Heroku, so I'm happily considering options.
It's clear this is not getting the level of attention I would expect for something so critical to operation (this ticket was opened 3 years ago).

**Comment by cmdrkeene** [comment]
+1 to adding this ASAP -- it's not a "nice-to-have" it's critical for many apps. People will just move if the delays continue.

**Comment by ezekg** [comment]
> Can anyone start listing alternative solutions here? At least it would be more helpful to people reading. We are on Standard 7 and I'm not going into another 4h maintenance. We decided to move away from Heroku, so I'm happily considering options. It's clear this is not getting the level of attention I would expect for something so critical to operation (this ticket was opened 3 years ago).

Good comparison across the alternative providers: https://planetscale.com/blog/benchmarking-postgres

Currently writing ~1GB/hr, which is eventually pruned but not reclaimed. This would be nice.

**Comment by jchatel** [comment]
How timely

```
We detected that your database (CRIMSON on xxx) is over capacity for your plan.

The standard-7 plan is intended for databases up to 2048 GB, but your database is using 2048 GB. You can view your current usage from your [database's dashboard](https://data.heroku.com/).
```

**Comment by henriquecf** [comment]
> Can anyone start listing alternative solutions here? At least it would be more helpful to people reading. We are on Standard 7 and I'm not going into another 4h maintenance. We decided to move away from Heroku, so I'm happily considering options. It's clear this is not getting the level of attention I would expect for something so critical to operation (this ticket was opened 3 years ago).

Hey we are also considering moving out of Heroku. Waiting on contract to be over. I would highly suggest considering PlanetScale for a better managed service. They also help migrating your DB with minimal downtime.

**Comment by ezekg** [comment]
Looking into PlanetScale, they also don't support `pg_repack`. But they do support [`pg_squeeze`](https://github.com/cybertec-postgresql/pg_squeeze/), which seems to be a replacement for `pg_repack`. Might be worth noting if Heroku is going to offer this as part of Postgres Advanced this year.

**Comment by swanson** [comment]
Would love to see this added, seems overdue. We currently are using our once-a-year scheduled holiday downtime to do vacuums on large tables...really not ideal.

**Comment by montulli** [comment]
We will likely have to leave Heroku unless this becomes available.   The performance gains of pg_repack are tremendous; it is a must-have feature!

**Comment by raoulalwani** [comment]
Checking in on this @jbrown-heroku and anyone from the Heroku team - any updates on this and the new platform? We honestly don't want to explore migrations, but will need to do so if there's continued silence on roadmap for this. For others, if you've migrated or are planning to migrate, what have you looked at?

**Comment by cyril-sf** [comment]
`Heroku is transitioning to a sustaining engineering model focused on stability, security, reliability, and support. Heroku remains an actively supported, production-ready platform, with an emphasis on maintaining quality and operational excellence rather than introducing new features.
`

It's really time to consider alternatives...

**Comment by patrick-boatner** [comment]
Heroku, you're out of touch with what your customers want. Instead of adding support for a simple open source extension (pg_repack), you built https://www.heroku.com/postgres-advanced-pilot/. I don't want your "innovative Follower Pool for zero-downtime scaling"; I just want a simple tool to cluster/shard my database.

**Comment by jnelson15** [comment]
+1

**Comment by sealabcore** [comment]
We really need pg_repack, any update here?

**Comment by cyril-sf** [comment]
https://www.heroku.com/blog/an-update-on-heroku/

> An Update on Heroku
> February 06, 2026 
> 
> Today, Heroku is transitioning to a sustaining engineering model focused on stability, security, reliability, and support. Heroku remains an actively supported, production-ready platform, with an emphasis on maintaining quality and operational excellence rather than introducing new features.
> 

Heroku doesn't want us as long-term customers, time to move on.

**Comment by sealabcore** [comment]
> https://www.heroku.com/blog/an-update-on-heroku/
> 
> > An Update on Heroku
> > February 06, 2026
> > Today, Heroku is transitioning to a sustaining engineering model focused on stability, security, reliability, and support. Heroku remains an actively supported, production-ready platform, with an emphasis on maintaining quality and operational excellence rather than introducing new features.
> 
> Heroku doesn't want us as long-term customers, time to move on.

Being able to maintain our database with pg_repack seems pretty important for quality and operational excellence.

**Comment by friism** [comment]
We appreciate your feedback on the importance of pg_repack support! As per our [blog](https://www.heroku.com/blog/march-2026-update/) we'll continue to release features that align with our sustaining engineering goals and your input helps with prioritization. And we are investigating ways to add pg_repack support, stay tuned.

**Comment by jessehanley** [comment]
@friism where is the new Heroku Postgres platform that was promised to be launched last year?

**Comment by cyril-sf** [comment]
https://www.techradar.com/pro/salesforce-halts-development-of-new-features-for-heroku-cloud-ai-platform

> Salesforce SVP and GM confirms Heroku gets deprioritized in an AI-first world
> Agentforce has emerged as an agentic AI development platform

https://www.techtarget.com/searchcustomerexperience/opinion/Salesforce-layoffs-continue-as-high-ranking-executives-leave

> Published: 11 Feb 2026
>
> Like many tech companies, Salesforce is laying off employees by the thousands; another roughly 1,000 were let go this week, with many outlets reporting that across-the-board reductions are hitting marketing, product management, analytics and Agentforce teams.
> Also affected in the layoffs was the Heroku platform as a service team. Salesforce said this week that Heroku will be sunsetted -- no new features will be developed, no new Enterprise Account contracts will be signed, but current customers will be able to continue using Heroku.

https://www.salesforceben.com/why-has-salesforce-stock-dropped-25-in-one-year/

> Salesforce’s stock – one of the company’s most widely talked-about topics of conversation – is something that will continue to be watched closely, especially as it has dropped 25.40% in just one year. Even after the launch of the AI tool Agentforce, multiple acquisitions, and growing financial results, the scales refuse to tilt up.

It starts to be hard to believe anything shared by the Heroku team here. With all the layoffs, I don't blame them. I just don't think they know or are in control, if anyone is at this point.

**Comment by sealabcore** [comment]
@friism looks like Postgres 19 has the feature natively built in. Do you know when Heroku will offer Postgres 19?

https://www.bytebase.com/blog/postgres-19-features-im-excited-about/

`REPACK TABLE my_table CONCURRENTLY;`

**Comment by wmcbroomd2d** [comment]
> [@friism](https://github.com/friism) looks like Postgres 19 has the feature natively built in. Do you know when Heroku will offer Postgres 19?
> 
> https://www.bytebase.com/blog/postgres-19-features-im-excited-about/
> 
> `REPACK TABLE my_table CONCURRENTLY;`


Postgres 19 is still in beta and isn't expected to be released until September.   

> The next major release of PostgreSQL is planned to be the 19 release. This release is planned for September 2026. 
https://www.postgresql.org/developer/roadmap/

**Comment by friism** [comment]
@jessehanley @cyril-sf postgres advanced is available in limited GA now: https://devcenter.heroku.com/changelog-items/3696

Send me an email at friis@heroku.com if you want to try it out.

**Comment by chaadow** [comment]
@friism hi, will pg_repack be part of postgres 18 release?
